datafusion/optimizer/src/optimizer.rs

```diff
@@ -445,35 +462,38 @@ impl Optimizer {
             }
             i += 1;
         }
+
+        // verify LP is valid, after the last optimizer pass.
+        new_plan
+            .check_invariants(InvariantLevel::Executable)
+            .map_err(|e| {
+                DataFusionError::Context(
+                    "check_plan_after_optimizers".to_string(),
+                    Box::new(e),
+                )
+            })?;
+
         log_plan("Final optimized plan", &new_plan);
         debug!("Optimizer took {} ms", start_time.elapsed().as_millis());
         Ok(new_plan)
     }
 }
 
-/// Returns an error if `new_plan`'s schema is different than `prev_schema`
+/// These are invariants which should hold true before and after each optimization.
 ///
-/// It ignores metadata and nullability.
-pub(crate) fn assert_schema_is_the_same(
+/// This differs from [`LogicalPlan::check_invariants`], which addresses if a singular
+/// LogicalPlan is valid. Instead this address if the optimization (before and after)
+/// is valid based upon permitted changes.
+fn assert_valid_optimization(
```

Comment:
Since all this does is call `assert_expected_schema `I think it might be easier to understand if you changed the callsite to call `assert_expected_schema` directly

Reply:
The purpose of `assert_valid_optimization` is define what is allowed to change in the optimizer passes. As of right now, we only defined that that schema could not change. The idea was that [future work](https://github.com/apache/datafusion/issues/13652#issuecomment-2550443777) could add to this method.

Reply:
Make sense